### PR TITLE
Repodata API

### PR DESCRIPTION
**Pull request by schuylermartin45**
@akabanovs @cbouss Adding you both for visibility. You two shouldn't feel required to review this one, unless you want to.

First pass at a new Repodata API that is type-safe and validates the JSON we pull in.

Also allows for easier version-ordering comparisons between two packages. This uses the `VersionOrder` class from `conda` and ALSO compares against the `build_number` we pulled from the `repodata.json` file.

### CODE REVIEW

File: anaconda_packaging_utils/api/repodata_api.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+    ## Required fields ##
+    build: str
+    build_number: int
+    depends: list[str]
+    md5: str
+    sha256: str
+    name: str
+    size: int
+    version: str
```
Comment (markan) [inline anaconda_packaging_utils/api/repodata_api.py:181]: I appreciate that you are deferring parsing the version here.

**Review by markan — COMMENTED** [review]
A few quick comments, I'll take time to read this in more detail later. 
This generally looks like a good starting point, but we may want to think about common access methods and indices.

File: anaconda_packaging_utils/api/repodata_api.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    timestamp: Optional[int] = None
+    date: Optional[str] = None
+    track_features: Optional[str] = None
+    license: Optional[str] = None
+    license_family: Optional[str] = None
+
+    # TODO rm: Enforce type checking when this PR is released
+    #   https://github.com/conda/conda/pull/13385
+    @no_type_check
+    def __lt__(self, other: VersionOrder) -> bool:
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply anaconda_packaging_utils/api/repodata_api.py:193]: That should actually be `object`, to be in alignment with "best practices": https://stackoverflow.com/questions/37557411/why-does-defining-the-argument-types-for-eq-throw-a-mypy-type-error

[the remark this answers is not in the document]
Comment (schuylermartin45) [reply anaconda_packaging_utils/api/repodata_api.py:193]: This is what happens when I use `@no_type_check` lol

File: anaconda_packaging_utils/api/repodata_api.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    OSX_X86_64 = "osx-64"
+    OSX_X86_32 = "osx-32"
+    WIN_64 = "win-64"
+    WIN_32 = "win-32"
+    NO_ARCH = "noarch"
+
+
+_BASE_REPODATA_URL: Final[str] = "https://repo.anaconda.com/pkgs"
+
+# Maps the available architectures per channel hosted on `repo.anaconda.com`
```
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply anaconda_packaging_utils/api/repodata_api.py:65]: I didn't realize that file existed and didn't want to write a HTML parser on the landing page. Good to know, I will gladly do that.

File: anaconda_packaging_utils/api/repodata_api.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    Structure that contains an entire serialized `repodata.json` blob
+    """
+
+    info: RepodataMetadata
+    packages: dict[str, PackageData]
+    removed: list[str]
+    repodata_version: int
+
+
+_REPODATA_JSON_SCHEMA: Final[SchemaType] = {
```
Comment (markan) [inline anaconda_packaging_utils/api/repodata_api.py:232]: Is this derived from https://github.com/conda/schemas?

File: anaconda_packaging_utils/api/repodata_api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (schuylermartin45) [reply anaconda_packaging_utils/api/repodata_api.py]: Done, pulled from the applicable `channeldata.json` file

File: anaconda_packaging_utils/api/repodata_api.py (hunk 4 above)
Comment (schuylermartin45) [reply anaconda_packaging_utils/api/repodata_api.py:164]: Don't know how I missed this comment initially. No it isn't, I didn't know this existed. I reverse engineered this schema from examples and then ran an automated test against multiple real world examples.

There is a new schema in the latest code for `channeldata.json` but I only validate the `subdirs` field I use to look-up supported architectures.

File: anaconda_packaging_utils/api/repodata_api.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+                return self.build_number < other.build_number
+            return self_v < other_v
+        except InvalidVersionSpec:
+            # If a `VersionOrder` object failed to be created, we are not able to accurately compare these objects.
+            pass
+        return False
+
+
+@dataclass
+class Repodata:
```
Comment (akabanovs) [inline anaconda_packaging_utils/api/repodata_api.py:153]: So this seems to be something similar to the Repodata class in anaconda-channel-scanner. The only difference is that there are more checks in this implementation, and that it processes all the info. Anaconda-channel-scanner just reads what it needs for the dependency analysis and also structures the package data into  a few levels: package-name/python-version/package-version/build-number.

Comment (schuylermartin45) [reply anaconda_packaging_utils/api/repodata_api.py:153]: Good to know. For Mark's needs (and I think percy's as well), all the data needs to be available at once. I wanted to create some more intelligent caching setup, but it also sounds like Cheng is on the warpath to replace these unscalable JSON files with something else.

So this is more of a stop-gap until we know what the "new and improved" version will be.

**Pull request merged**